### PR TITLE
Threads and openorbis stuff

### PR DESCRIPTION
**Pull request by georgemoralis**
some rework on HLE threads and trying to work on getting openorbis work again with these

### CODE REVIEW

File: src/core/hle/libraries/libkernel/libkernel.cpp [hunk 1]
```diff
@@ -28,11 +32,79 @@ int32_t PS4_SYSV_ABI sceKernelReleaseDirectMemory(off_t start, size_t len) {
...
+    std::this_thread::sleep_for(std::chrono::microseconds(microseconds));
+}
+
+struct iovec {
+    void* iov_base; /* Base	address. */
+    size_t iov_len; /* Length. */
+};
+
+size_t PS4_SYSV_ABI _writev(int fd, const struct iovec* iov, int iovcn) {
+    //weird it gives fd ==0 and writes to stdout , i am not sure if it that is valid (found in openorbis)
```
Comment (georgemoralis) [inline src/core/hle/libraries/libkernel/libkernel.cpp:50]: ok this is weird...
openorbis hello world says that fd==0 goes to stdout but that should not be normal

File: src/core/linker.cpp [hunk 2]
```diff
@@ -622,6 +623,31 @@ void Linker::Resolve(const std::string& name, int Symtype, Module* m, Loader::Sy
 
 }
 
+static thread_local void* tls_block_top;
+
+// this assumes TLS will do mov rax|rcx, fs:[0] and
+// that fs:[0] will be a nullptr in windows (seems to be so)
+// and, that fs:[0] points to the top of the static TLS
+static LONG CALLBACK ExceptionHandlerTLS(PEXCEPTION_POINTERS ExceptionInfo) {
```
Comment (georgemoralis) [inline src/core/linker.cpp:631]: ok that is temp will surely be rewritten

File: src/core/linker.cpp [hunk 3]
```diff
@@ -632,6 +658,11 @@ static PS4_SYSV_ABI void ProgramExitFunc() {
 static void run_main_entry(u64 addr, EntryParams* params, exit_func_t exit_func) {
     //reinterpret_cast<entry_func_t>(addr)(params, exit_func); // can't be used, stack has to have a specific layout
 
+	void** tls_block = (void**)malloc(16);  // this assumes 8 bytes of TLS storage, with the top of TLS pointing to itself
```
Comment (georgemoralis) [inline src/core/linker.cpp:661]: temp hackish code

File: src/core/hle/libraries/libkernel/thread_management.cpp [hunk 4]
```diff
@@ -8,22 +8,28 @@
 
 namespace Core::Libraries::LibKernel {
 
-thread_local PthreadInternal g_pthread_self{};
+thread_local ScePthread g_pthread_self{};
```
Comment (raphaelthegreat) [inline src/core/hle/libraries/libkernel/thread_management.cpp:11]: Does this need to be a pointer? The old way avoided a call to new which you aren't freeing.

File: src/core/hle/libraries/libkernel/thread_management.cpp [hunk 5]
```diff
@@ -275,26 +385,47 @@ int PS4_SYSV_ABI posix_pthread_mutex_unlock(ScePthreadMutex* mutex) {
     return result;
 }
 
+int PS4_SYSV_ABI posix_pthread_cond_broadcast(ScePthreadCond* cond) {
+    LOG_INFO_IF(log_pthread_file, "posix posix_pthread_cond_broadcast redirect to scePthreadCondBroadcast\n");
+    int result = scePthreadCondBroadcast(cond);
+    if (result != 0) {
+        int rt = result > SCE_KERNEL_ERROR_UNKNOWN && result <= SCE_KERNEL_ERROR_ESTOP ? result + -SCE_KERNEL_ERROR_UNKNOWN : POSIX_EOTHER;
```
Comment (raphaelthegreat) [inline src/core/hle/libraries/libkernel/thread_management.cpp:392]: What does result + -SCE_KERNEL_ERROR_UNKNOWN mean?

File: src/core/hle/libraries/libkernel/libkernel.cpp [hunk 6]
```diff
@@ -28,11 +32,79 @@ int32_t PS4_SYSV_ABI sceKernelReleaseDirectMemory(off_t start, size_t len) {
...
+#define PROT_READ 0x1
+#define PROT_WRITE 0x2
+
+int PS4_SYSV_ABI sceKernelMmap(void* addr, u64 len, int prot, int flags, int fd, off_t offset, void** res) {
+    PRINT_FUNCTION_NAME();
+    if (prot > 3)// READ,WRITE or bitwise READ | WRITE supported
+    {
+        LOG_ERROR_IF(log_libkernel_file, "sceKernelMmap prot ={} not supported\n",prot);
+    }
+    DWORD flProtect;
```
Comment (raphaelthegreat) [inline src/core/hle/libraries/libkernel/libkernel.cpp:70]: If PROT_WRITE is not specified, this will have a garbage value. Initialize it appropriately

Comment (georgemoralis) [reply src/core/hle/libraries/libkernel/libkernel.cpp:70]: hmm?

**Comment by georgemoralis** [comment]
absoulute

**Pull request closed without merging**